In v3.1.0 netutil unit test failed on ipv6only linux host

Hello!

On last release github.com/coreos/etcd/pkg/netutil unit test was failed if ipv4 adress is missed on host.
Simple script to reproduce:

```(bash)
wget https://github.com/coreos/etcd/archive/v3.1.0.zip
wget https://storage.googleapis.com/golang/go1.7.4.linux-amd64.tar.gz
unzip v3.1.0.zip
tar xf go1.7.4.linux-amd64.tar.gz
GOROOT=$(pwd)/go
PATH=$GOROOT/bin:$PATH
cd etcd-3.1.0/
bash test
# all success
# then let's delete ipv4
sudo ip a delete dev enp0s31f6 $MY_IPV4
# device look like
ip a l dev enp0s31f6 
3: enp0s31f6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 84:7b:eb:07:ed:92 brd ff:ff:ff:ff:ff:ff
    inet6 a:b:c:401:867b:ebff:fe07:ed92/64 scope global noprefixroute dynamic 
       valid_lft 115198sec preferred_lft 57598sec
    inet6 fe80::867b:ebff:fe07:ed92/64 scope link 
       valid_lft forever preferred_lft forever
# try test again
bash test
```
The output from the last command:
```Running unit tests...
ok  	github.com/coreos/etcd/auth	2.423s	coverage: 46.8% of statements
ok  	github.com/coreos/etcd/client	1.622s	coverage: 43.6% of statements
ok  	github.com/coreos/etcd/clientv3	8.075s	coverage: 24.3% of statements
ok  	github.com/coreos/etcd/clientv3/naming	1.850s	coverage: 78.3% of statements
ok  	github.com/coreos/etcd/compactor	3.430s	coverage: 89.7% of statements
ok  	github.com/coreos/etcd/discovery	1.397s	coverage: 65.7% of statements
ok  	github.com/coreos/etcd/embed	1.042s	coverage: 16.6% of statements
ok  	github.com/coreos/etcd/error	1.025s	coverage: 84.6% of statements
ok  	github.com/coreos/etcd/etcdctl/ctlv2/command	1.051s	coverage: 0.7% of statements
ok  	github.com/coreos/etcd/etcdmain	1.140s	coverage: 28.2% of statements
ok  	github.com/coreos/etcd/etcdserver	4.671s	coverage: 28.2% of statements
ok  	github.com/coreos/etcd/etcdserver/api/v2http	1.105s	coverage: 72.9% of statements
ok  	github.com/coreos/etcd/etcdserver/api/v2http/httptypes	1.014s	coverage: 76.2% of statements
ok  	github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes	1.017s	coverage: 75.0% of statements
ok  	github.com/coreos/etcd/etcdserver/auth	4.344s	coverage: 75.1% of statements
ok  	github.com/coreos/etcd/etcdserver/membership	1.060s	coverage: 59.7% of statements
ok  	github.com/coreos/etcd/lease	7.360s	coverage: 77.2% of statements
ok  	github.com/coreos/etcd/lease/leasehttp	7.081s	coverage: 65.1% of statements
ok  	github.com/coreos/etcd/mvcc	9.911s	coverage: 78.0% of statements
ok  	github.com/coreos/etcd/mvcc/backend	3.012s	coverage: 82.1% of statements
ok  	github.com/coreos/etcd/pkg/adt	2.092s	coverage: 88.1% of statements
ok  	github.com/coreos/etcd/pkg/cors	1.012s	coverage: 92.9% of statements
ok  	github.com/coreos/etcd/pkg/crc	1.008s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/expect	1.021s	coverage: 93.9% of statements
ok  	github.com/coreos/etcd/pkg/fileutil	1.450s	coverage: 61.1% of statements
ok  	github.com/coreos/etcd/pkg/flags	1.011s	coverage: 57.5% of statements
ok  	github.com/coreos/etcd/pkg/idutil	1.009s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/ioutil	1.080s	coverage: 74.6% of statements
ok  	github.com/coreos/etcd/pkg/logutil	7.158s	coverage: 58.5% of statements
ok  	github.com/coreos/etcd/pkg/monotime	1.016s	coverage: 50.0% of statements
2017-01-23 16:20:41.591931 I | pkg/netutil: resolving infra0.example.com:4001 to 10.0.1.10:4001
2017-01-23 16:20:41.592117 I | pkg/netutil: resolving infra0.example.com:2379 to 10.0.1.10:2379
2017-01-23 16:20:41.592196 I | pkg/netutil: resolving infra0.example.com:7001 to 10.0.1.10:7001
2017-01-23 16:20:41.592235 I | pkg/netutil: resolving infra0.example.com:2380 to 10.0.1.10:2380
2017-01-23 16:20:41.592348 W | pkg/netutil: failed resolving host infra0.example.com:4001 (cannot resolve host.); retrying in 1s
2017-01-23 16:20:42.594086 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.594560 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.594749 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.595097 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.595424 I | pkg/netutil: resolving example.com:2380 to 10.0.10.1:2380
2017-01-23 16:20:42.595675 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.596032 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.596572 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:42.597018 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
2017-01-23 16:20:42.597206 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:42.597535 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:42.597679 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
--- FAIL: TestGetDefaultInterface (0.00s)
	routes_linux_test.go:24: could not find default interface
2017-01-23 16:20:42.601435 I | pkg/netutil: resolving infra0.example.com:4001 to 10.0.1.10:4001
2017-01-23 16:20:42.601596 I | pkg/netutil: resolving infra0.example.com:2379 to 10.0.1.10:2379
2017-01-23 16:20:42.601787 I | pkg/netutil: resolving infra0.example.com:7001 to 10.0.1.10:7001
2017-01-23 16:20:42.601913 I | pkg/netutil: resolving infra0.example.com:2380 to 10.0.1.10:2380
2017-01-23 16:20:42.602250 W | pkg/netutil: failed resolving host infra0.example.com:4001 (cannot resolve host.); retrying in 1s
2017-01-23 16:20:43.602734 E | pkg/netutil: could not resolve host infra0.example.com:4001
2017-01-23 16:20:43.603966 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.604516 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.604758 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.605147 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.605509 I | pkg/netutil: resolving example.com:2380 to 10.0.10.1:2380
2017-01-23 16:20:43.605818 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.606294 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.606807 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:43.607151 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
2017-01-23 16:20:43.607313 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:43.607681 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:43.607847 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
--- FAIL: TestGetDefaultInterface (0.00s)
	routes_linux_test.go:24: could not find default interface
2017-01-23 16:20:43.612154 I | pkg/netutil: resolving infra0.example.com:4001 to 10.0.1.10:4001
2017-01-23 16:20:43.612346 I | pkg/netutil: resolving infra0.example.com:2379 to 10.0.1.10:2379
2017-01-23 16:20:43.612604 I | pkg/netutil: resolving infra0.example.com:7001 to 10.0.1.10:7001
2017-01-23 16:20:43.612768 I | pkg/netutil: resolving infra0.example.com:2380 to 10.0.1.10:2380
2017-01-23 16:20:43.613214 W | pkg/netutil: failed resolving host infra0.example.com:4001 (cannot resolve host.); retrying in 1s
2017-01-23 16:20:44.613401 E | pkg/netutil: could not resolve host infra0.example.com:4001
2017-01-23 16:20:44.613976 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.614220 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.614317 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.614493 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.614673 I | pkg/netutil: resolving example.com:2380 to 10.0.10.1:2380
2017-01-23 16:20:44.614823 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.615031 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.615274 I | pkg/netutil: resolving example.com:2379 to 10.0.10.1:2379
2017-01-23 16:20:44.615442 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
2017-01-23 16:20:44.615513 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:44.615684 I | pkg/netutil: resolving second.com:2380 to 10.0.11.2:2380
2017-01-23 16:20:44.615777 I | pkg/netutil: resolving first.com:2379 to 10.0.11.1:2379
--- FAIL: TestGetDefaultInterface (0.00s)
	routes_linux_test.go:24: could not find default interface
FAIL
coverage: 56.1% of statements
FAIL	github.com/coreos/etcd/pkg/netutil	3.033s
ok  	github.com/coreos/etcd/pkg/osutil	1.014s	coverage: 87.1% of statements
ok  	github.com/coreos/etcd/pkg/pathutil	1.007s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/pbutil	1.011s	coverage: 76.9% of statements
ok  	github.com/coreos/etcd/pkg/report	1.008s	coverage: 24.1% of statements
ok  	github.com/coreos/etcd/pkg/schedule	1.011s	coverage: 85.7% of statements
ok  	github.com/coreos/etcd/pkg/testutil	1.071s	coverage: 20.2% of statements
ok  	github.com/coreos/etcd/pkg/transport	1.554s	coverage: 66.8% of statements
ok  	github.com/coreos/etcd/pkg/types	1.027s	coverage: 90.8% of statements
ok  	github.com/coreos/etcd/pkg/wait	1.167s	coverage: 92.9% of statements
ok  	github.com/coreos/etcd/proxy/grpcproxy	1.566s	coverage: 5.9% of statements
ok  	github.com/coreos/etcd/proxy/httpproxy	1.020s	coverage: 75.8% of statements
ok  	github.com/coreos/etcd/proxy/tcpproxy	1.026s	coverage: 64.2% of statements
ok  	github.com/coreos/etcd/raft	5.833s	coverage: 86.8% of statements
ok  	github.com/coreos/etcd/rafthttp	2.444s	coverage: 75.8% of statements
ok  	github.com/coreos/etcd/snap	1.029s	coverage: 54.2% of statements
ok  	github.com/coreos/etcd/store	1.205s	coverage: 90.8% of statements
ok  	github.com/coreos/etcd/wal	4.844s	coverage: 80.5% of statements
```

--- a/embed/config.go
+++ b/embed/config.go
@@ -68,8 +68,8 @@ func init() {
 		return
 	}
 	// found default host, advertise on it
-	DefaultInitialAdvertisePeerURLs = "http://" + ip + ":2380"
-	DefaultAdvertiseClientURLs = "http://" + ip + ":2379"
+	DefaultInitialAdvertisePeerURLs = "http://" + net.JoinHostPort(ip, "2380")
+	DefaultAdvertiseClientURLs = "http://" + net.JoinHostPort(ip, "2379")
 	defaultHostname = ip
 }
 

--- a/pkg/netutil/isolate_linux.go
+++ b/pkg/netutil/isolate_linux.go
@@ -43,7 +43,7 @@ func RecoverPort(port int) error {
 
 // SetLatency adds latency in millisecond scale with random variations.
 func SetLatency(ms, rv int) error {
-	ifce, err := GetDefaultInterface()
+	ifces, err := GetDefaultInterfaces()
 	if err != nil {
 		return err
 	}
@@ -51,14 +51,16 @@ func SetLatency(ms, rv int) error {
 	if rv > ms {
 		rv = 1
 	}
-	cmdStr := fmt.Sprintf("sudo tc qdisc add dev %s root netem delay %dms %dms distribution normal", ifce, ms, rv)
-	_, err = exec.Command("/bin/sh", "-c", cmdStr).Output()
-	if err != nil {
-		// the rule has already been added. Overwrite it.
-		cmdStr = fmt.Sprintf("sudo tc qdisc change dev %s root netem delay %dms %dms distribution normal", ifce, ms, rv)
+	for ifce := range ifces {
+		cmdStr := fmt.Sprintf("sudo tc qdisc add dev %s root netem delay %dms %dms distribution normal", ifce, ms, rv)
 		_, err = exec.Command("/bin/sh", "-c", cmdStr).Output()
 		if err != nil {
-			return err
+			// the rule has already been added. Overwrite it.
+			cmdStr = fmt.Sprintf("sudo tc qdisc change dev %s root netem delay %dms %dms distribution normal", ifce, ms, rv)
+			_, err = exec.Command("/bin/sh", "-c", cmdStr).Output()
+			if err != nil {
+				return err
+			}
 		}
 	}
 	return nil
@@ -66,10 +68,15 @@ func SetLatency(ms, rv int) error {
 
 // RemoveLatency resets latency configurations.
 func RemoveLatency() error {
-	ifce, err := GetDefaultInterface()
+	ifces, err := GetDefaultInterfaces()
 	if err != nil {
 		return err
 	}
-	_, err = exec.Command("/bin/sh", "-c", fmt.Sprintf("sudo tc qdisc del dev %s root netem", ifce)).Output()
-	return err
+	for ifce := range ifces {
+		_, err = exec.Command("/bin/sh", "-c", fmt.Sprintf("sudo tc qdisc del dev %s root netem", ifce)).Output()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }

--- a/pkg/netutil/routes_linux.go
+++ b/pkg/netutil/routes_linux.go
@@ -27,42 +27,49 @@ import (
 )
 
 var errNoDefaultRoute = fmt.Errorf("could not find default route")
+var errNoDefaultHost = fmt.Errorf("could not find default host")
+var errNoDefaultInterface = fmt.Errorf("could not find default interface")
 
+// GetDefaultHost obtains the first IP address of machine from the routing table and returns the IP address as string.
+// An IPv4 address is preferred to an IPv6 address for backward compatibility.
 func GetDefaultHost() (string, error) {
-	rmsg, rerr := getDefaultRoute()
+	rmsgs, rerr := getDefaultRoutes()
 	if rerr != nil {
 		return "", rerr
 	}
 
-	host, oif, err := parsePREFSRC(rmsg)
-	if err != nil {
-		return "", err
-	}
-	if host != "" {
-		return host, nil
-	}
+	for family, rmsg := range rmsgs {
+		host, oif, err := parsePREFSRC(rmsg)
+		if err != nil {
+			return "", err
+		}
+		if host != "" {
+			return host, nil
+		}
 
-	// prefsrc not detected, fall back to getting address from iface
-	ifmsg, ierr := getIface(oif)
-	if ierr != nil {
-		return "", ierr
-	}
+		// prefsrc not detected, fall back to getting address from iface
+		ifmsg, ierr := getIfaceAddr(oif, family)
+		if ierr != nil {
+			return "", ierr
+		}
 
-	attrs, aerr := syscall.ParseNetlinkRouteAttr(ifmsg)
-	if aerr != nil {
-		return "", aerr
-	}
+		attrs, aerr := syscall.ParseNetlinkRouteAttr(ifmsg)
+		if aerr != nil {
+			return "", aerr
+		}
 
-	for _, attr := range attrs {
-		if attr.Attr.Type == syscall.RTA_SRC {
-			return net.IP(attr.Value).String(), nil
+		for _, attr := range attrs {
+			// search for RTA_DST because ipv6 doesn't have RTA_SRC
+			if attr.Attr.Type == syscall.RTA_DST {
+				return net.IP(attr.Value).String(), nil
+			}
 		}
 	}
 
-	return "", errNoDefaultRoute
+	return "", errNoDefaultHost
 }
 
-func getDefaultRoute() (*syscall.NetlinkMessage, error) {
+func getDefaultRoutes() (map[uint8]*syscall.NetlinkMessage, error) {
 	dat, err := syscall.NetlinkRIB(syscall.RTM_GETROUTE, syscall.AF_UNSPEC)
 	if err != nil {
 		return nil, err
@@ -73,6 +80,7 @@ func getDefaultRoute() (*syscall.NetlinkMessage, error) {
 		return nil, msgErr
 	}
 
+	routes := make(map[uint8]*syscall.NetlinkMessage)
 	rtmsg := syscall.RtMsg{}
 	for _, m := range msgs {
 		if m.Header.Type != syscall.RTM_NEWROUTE {
@@ -82,17 +90,23 @@ func getDefaultRoute() (*syscall.NetlinkMessage, error) {
 		if rerr := binary.Read(buf, cpuutil.ByteOrder(), &rtmsg); rerr != nil {
 			continue
 		}
-		if rtmsg.Dst_len == 0 {
+		if rtmsg.Dst_len == 0 && rtmsg.Table == syscall.RT_TABLE_MAIN {
 			// zero-length Dst_len implies default route
-			return &m, nil
+			msg := m
+			routes[rtmsg.Family] = &msg
 		}
+	}
+
+	if len(routes) > 0 {
+		return routes, nil
 	}
 
 	return nil, errNoDefaultRoute
 }
 
-func getIface(idx uint32) (*syscall.NetlinkMessage, error) {
-	dat, err := syscall.NetlinkRIB(syscall.RTM_GETADDR, syscall.AF_UNSPEC)
+// Used to get an address of interface.
+func getIfaceAddr(idx uint32, family uint8) (*syscall.NetlinkMessage, error) {
+	dat, err := syscall.NetlinkRIB(syscall.RTM_GETADDR, int(family))
 	if err != nil {
 		return nil, err
 	}
@@ -116,38 +130,75 @@ func getIface(idx uint32) (*syscall.NetlinkMessage, error) {
 		}
 	}
 
-	return nil, errNoDefaultRoute
+	return nil, fmt.Errorf("could not find address for interface index %v", idx)
+
 }
 
-var errNoDefaultInterface = fmt.Errorf("could not find default interface")
-
-func GetDefaultInterface() (string, error) {
-	rmsg, rerr := getDefaultRoute()
-	if rerr != nil {
-		return "", rerr
-	}
-
-	_, oif, err := parsePREFSRC(rmsg)
+// Used to get a name of interface.
+func getIfaceLink(idx uint32) (*syscall.NetlinkMessage, error) {
+	dat, err := syscall.NetlinkRIB(syscall.RTM_GETLINK, syscall.AF_UNSPEC)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 
-	ifmsg, ierr := getIface(oif)
-	if ierr != nil {
-		return "", ierr
+	msgs, msgErr := syscall.ParseNetlinkMessage(dat)
+	if msgErr != nil {
+		return nil, msgErr
 	}
 
-	attrs, aerr := syscall.ParseNetlinkRouteAttr(ifmsg)
-	if aerr != nil {
-		return "", aerr
-	}
-
-	for _, attr := range attrs {
-		if attr.Attr.Type == syscall.IFLA_IFNAME {
-			return string(attr.Value[:len(attr.Value)-1]), nil
+	ifinfomsg := syscall.IfInfomsg{}
+	for _, m := range msgs {
+		if m.Header.Type != syscall.RTM_NEWLINK {
+			continue
+		}
+		buf := bytes.NewBuffer(m.Data[:syscall.SizeofIfInfomsg])
+		if rerr := binary.Read(buf, cpuutil.ByteOrder(), &ifinfomsg); rerr != nil {
+			continue
+		}
+		if ifinfomsg.Index == int32(idx) {
+			return &m, nil
 		}
 	}
-	return "", errNoDefaultInterface
+
+	return nil, fmt.Errorf("could not find link for interface index %v", idx)
+}
+
+// GetDefaultInterfaces gets names of interfaces and returns a map[interface]families.
+func GetDefaultInterfaces() (map[string]uint8, error) {
+	interfaces := make(map[string]uint8)
+	rmsgs, rerr := getDefaultRoutes()
+	if rerr != nil {
+		return interfaces, rerr
+	}
+
+	for family, rmsg := range rmsgs {
+		_, oif, err := parsePREFSRC(rmsg)
+		if err != nil {
+			return interfaces, err
+		}
+
+		ifmsg, ierr := getIfaceLink(oif)
+		if ierr != nil {
+			return interfaces, ierr
+		}
+
+		attrs, aerr := syscall.ParseNetlinkRouteAttr(ifmsg)
+		if aerr != nil {
+			return interfaces, aerr
+		}
+
+		for _, attr := range attrs {
+			if attr.Attr.Type == syscall.IFLA_IFNAME {
+				// key is an interface name
+				// possible values: 2 - AF_INET, 10 - AF_INET6, 12 - dualstack
+				interfaces[string(attr.Value[:len(attr.Value)-1])] += family
+			}
+		}
+	}
+	if len(interfaces) > 0 {
+		return interfaces, nil
+	}
+	return interfaces, errNoDefaultInterface
 }
 
 // parsePREFSRC returns preferred source address and output interface index (RTA_OIF).

--- a/pkg/netutil/routes_linux_test.go
+++ b/pkg/netutil/routes_linux_test.go
@@ -19,9 +19,17 @@ package netutil
 import "testing"
 
 func TestGetDefaultInterface(t *testing.T) {
-	ifc, err := GetDefaultInterface()
+	ifc, err := GetDefaultInterfaces()
 	if err != nil {
 		t.Fatal(err)
 	}
-	t.Logf("default network interface: %q\n", ifc)
+	t.Logf("default network interfaces: %+v\n", ifc)
+}
+
+func TestGetDefaultHost(t *testing.T) {
+	ip, err := GetDefaultHost()
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("default ip: %v", ip)
 }

